src/wasm/wasm-ir-builder.cpp

```diff
@@ -0,0 +1,703 @@
+    }
+  }
+
+  if (firstNone == 0) {
+    // There are no expressions that produce values.
+    if (unreachable) {
+      return builder.makeUnreachable();
+    }
+    return Err{"popping from empty stack"};
+  }
+
+  if (firstNone == exprStack.size()) {
+    // The last expression produced a value.
+    auto expr = exprStack.back();
+    exprStack.pop_back();
+    return expr;
+  }
+
+  // We need to assemble a block of expressions that returns the value of the
+  // first one using a scratch local (unless it's unreachable, in which case
+  // we can throw the following expressions away).
+  auto* expr = exprStack[firstNone - 1];
+  if (expr->type == Type::unreachable) {
+    exprStack.resize(firstNone - 1);
+    return expr;
+  }
+  auto scratchIdx = addScratchLocal(expr->type);
+  CHECK_ERR(scratchIdx);
+  std::vector<Expression*> exprs;
+  exprs.reserve(exprStack.size() - firstNone + 2);
+  exprs.push_back(builder.makeLocalSet(*scratchIdx, expr));
+  exprs.insert(exprs.end(), exprStack.begin() + firstNone, exprStack.end());
+  exprs.push_back(builder.makeLocalGet(*scratchIdx, expr->type));
+
+  exprStack.resize(firstNone - 1);
+  return builder.makeBlock(exprs, expr->type);
+}
+
+Expression* IRBuilder::build() {
+  auto& exprStack = getExprStack();
+  assert(scopeStack.size() == 1);
+  assert(exprStack.size() == 1);
+
+  auto e = exprStack.back();
+  exprStack.clear();
+  unreachable = false;
+  return e;
+}
+
+Result<std::vector<Expression*>> IRBuilder::finishInstrs() {
+  auto& exprStack = getExprStack();
+  auto type = getResultType();
+
+  // We have finished parsing a sequence of instructions. Fix up the parsed
+  // instructions and reset the context for the next sequence.
+  if (type.isTuple()) {
+    std::vector<Expression*> elems(type.size());
+    bool hadUnreachableElem = false;
+    for (size_t i = 0; i < elems.size(); ++i) {
+      auto elem = pop();
+      CHECK_ERR(elem);
+      elems[elems.size() - 1 - i] = *elem;
+      if ((*elem)->type == Type::unreachable) {
+        // We don't want to pop back past an unreachable here. Push the
+        // unreachable back and throw away any post-unreachable values we have
+        // popped.
+        exprStack.push_back(*elem);
+        hadUnreachableElem = true;
+        break;
+      }
+    }
+    if (!hadUnreachableElem) {
+      exprStack.push_back(builder.makeTupleMake(std::move(elems)));
+    }
+  } else if (type != Type::none) {
+    // Ensure the last expression produces the value.
+    auto expr = pop();
+    CHECK_ERR(expr);
+    exprStack.push_back(*expr);
+  }
+  unreachable = false;
+  auto ret = std::move(exprStack);
+  scopeStack.pop_back();
+  return ret;
+}
+
+Result<> IRBuilder::visit(Expression* curr) {
+  UnifiedExpressionVisitor<IRBuilder, Result<>>::visit(curr);
+  if (auto* block = curr->dynCast<Block>()) {
+    block->finalize(block->type);
+  } else {
+    // TODO: Call more efficient versions of finalize() for other kinds of nodes
+    // as well.
```

Comment:
```suggestion
    // as well (passing in the known type, as done above).
```
(it took me a few seconds to realize this)